app/services/conversation_bot.rb

```diff
@@ -1,13 +1,297 @@
 class ConversationBot
+  NOW_TIME = /now|ahora/i.freeze
 
+  CALL_FOR_HELP = /\Ahelp|ayuda\z/i.freeze
+
+  NUMBER_STRINGS = {
+      0 => [/0/, /zero/, /cero/],
+      1 => [/1/, /one/, /uno/],
+      2 => [/2/, /two/, /dos/],
+      3 => [/3/, /three/, /tres/],
+      4 => [/4/, /four/, /cuatro/]
+  }.freeze
+
+  DONT_KNOW_FRAGMENTS = [/don't know/i, /no se/i, /no sé/i].freeze
+
+  # accepts a conversation at a certain state and a new message that has arrived
   def initialize convo, message
     @conversation = convo
+    @bot_counter = convo.bot_counter
+    @username = @conversation.username.split(' ').first
     @message = message
+    @body = @message.body.strip
+    @locale = convo.user.language
   end
 
+  # returns an appropriate response to the new message and updates the conversation
+  # we have one method for each conversation lifecycle and it returns the retries
+  # count for the conversation. Updating this on the conversation will update the
+  # conversation lifecyle.
   def response
-    # todo - bot based on lifecycle
-    'Hi, thanks for contacting us. Someone will text you momentarily to arrange a ride.'
+    if @body =~ CALL_FOR_HELP
+      stalled
+    else
+      current_lifecycle = @conversation.lifecycle
+      new_counter = send(current_lifecycle.to_sym)
+      # force update if counter doesn't change
+      @conversation.update_attributes(bot_counter: new_counter, updated_at: Time.now)
+      @response
```

Comment:
@response is set by the method called in the send(). current_lifecycle is
one line above the send. Basically we are taking the lifecycle string
('have_name') converting it to a symbol and calling that method on this
object (def have_name). That method sets @response and returns the new bot
counter. I suppose it could have been a [response, new_counter] dual return
value. I go back and forth on returning multiple values.

On Mon, Aug 15, 2016 at 10:35 AM, John McGrath <notifications@github.com>
wrote:

> In app/services/conversation_bot.rb
> <https://github.com/DevProgress/drive.vote/pull/64#discussion_r74799614>:
>
> >    def response
> > -    # todo - bot based on lifecycle
> > -    'Hi, thanks for contacting us. Someone will text you momentarily to arrange a ride.'
> > +    if @body =~ CALL_FOR_HELP
> > +      stalled
> > +    else
> > +      current_lifecycle = @conversation.lifecycle
> > +      new_counter = send(current_lifecycle.to_sym)
> > +      # force update if counter doesn't change
> > +      @conversation.update_attributes(bot_counter: new_counter, updated_at: Time.now)
> > +      @response
>
> Where is @response <https://github.com/response> created? When
> current_lifecycle is set?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/DevProgress/drive.vote/pull/64/files/ddc41bfeefca580c93c6b12be6080d6e824f82ae#r74799614>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATwgOCLWkzHTp10e_Pz1EFmJeOUcZYtMks5qgKNUgaJpZM4Jj2K3>
> .
>